DualStack: basic dual-stack node IP handling

danwinship opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Makes two related improvements to kubelet node IP handling:

- On bare metal (ie, when there is no `--cloud-provider`), the administrator can now pass a comma-separated pair of IPs via `--node-ip` to manually configure dual-stack `node.status.addresses`. No other scenarios (eg, cloud provider, bare-metal autodetection) are changed.
- On nodes with dual-stack `node.status.addresses` (regardless of how they got them), hostNetwork pods will now have dual-stack `pod.status.podIPs`, allowing the pod to be the target of either a single-stack IPv4 or single-stack IPv6 Service (or, once #91824 merges, a dual-stack Service).

There is ongoing discussion about dual-stack node IPs in https://github.com/kubernetes/enhancements/pull/1665, and it probably won't be resolved in time for Enhancement Freeze. However, it seems like there has always been consensus that (a) regardless of what else is possible, it ought to be possible to explicitly specify a dual-stack pair of IPs on bare-metal nodes, (b) regardless of how a node ends up with dual-stack IPs, if it has dual-stack IPs, then that should be reflected in hostNetwork pods. So this PR implements those two things. Additional functionality will presumably come later.

**Which issue(s) this PR fixes**:
Fixes #91940
Fixes #89816 (invalid node IP set by cloud provider causes `"<nil>"` podIP on hostnetwork pod)

**Special notes for your reviewer**:
Currently built on top of #94474 (the first two commits). That can still be reviewed separately, or it can be reviewed as part of this.

Does not conflict with #91824 (dual-stack Services), as this PR only touches kubelet-related things, and that PR does not touch anything kubelet related except a small fix to kubenet (which this PR doesn't touch).

**Does this PR introduce a user-facing change?**:
```release-note
In dual-stack bare-metal clusters, you can now pass dual-stack IPs to `kubelet --node-ip`.
eg: `kubelet --node-ip 10.1.0.5,fd01::0005`. This is not yet supported for non-bare-metal
clusters.

In dual-stack clusters where nodes have dual-stack addresses, hostNetwork pods
will now get dual-stack PodIPs.
```

/sig network
/sig node
/priority important-soon
/assign @thockin

danwinship committed:
kubelet: allow specifying dual-stack node IPs on bare metal

Discussion is ongoing about how to best handle dual-stack with clouds
and autodetected IPs, but there is at least agreement that people on
bare metal ought to be able to specify two explicit IPs on dual-stack
hosts, so allow that.

danwinship committed:
kubelet: Set dual-stack hostNetwork pod IPs on dual-stack nodes

Add nodeutil.GetNodeHostIPs to return dual-stack node IPs (in
dual-stack clusters), and make kubelet use it.

danwinship committed:
utilnode: fix incorrect documentation about node name vs hostname

(The behavior is unchanged; the function always took a node name
rather than a hostname. Now it just documents that correctly.)

dcbw commented:
@danwinship verify failed:

```
Verifying verify-gofmt.sh

+++ Running case: verify.gofmt 
+++ working dir: /home/prow/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-gofmt.sh"
diff -u ./pkg/util/node/node_test.go.orig ./pkg/util/node/node_test.go
--- ./pkg/util/node/node_test.go.orig	2020-10-01 17:24:02.206547690 +0000
+++ ./pkg/util/node/node_test.go	2020-10-01 17:24:02.206547690 +0000
@@ -107,14 +107,14 @@
 			expectIPs: nil,
 		},
 		{
-			name:      "no InternalIP/ExternalIP",
+			name: "no InternalIP/ExternalIP",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeHostName, Address: "example.com"},
 			},
 			expectIPs: nil,
 		},
 		{
-			name:      "IPv4-only, simple",
+			name: "IPv4-only, simple",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
 				{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
@@ -123,7 +123,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4")},
 		},
 		{
-			name:      "IPv4-only, external-first",
+			name: "IPv4-only, external-first",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
 				{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
@@ -132,7 +132,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4")},
 		},
 		{
-			name:      "IPv4-only, no internal",
+			name: "IPv4-only, no internal",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
 				{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
@@ -140,7 +140,7 @@
 			expectIPs: []net.IP{net.ParseIP("4.3.2.1")},
 		},
 		{
-			name:      "dual-stack node, single-stack cluster",
+			name: "dual-stack node, single-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
 				{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
@@ -151,7 +151,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4")},
 		},
 		{
-			name:      "dual-stack node, dual-stack cluster",
+			name: "dual-stack node, dual-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
 				{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
@@ -163,7 +163,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4"), net.ParseIP("a:b::c:d")},
 		},
 		{
-			name:      "dual-stack node, different order, single-stack cluster",
+			name: "dual-stack node, different order, single-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
 				{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
@@ -174,7 +174,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4")},
 		},
 		{
-			name:      "dual-stack node, different order, dual-stack cluster",
+			name: "dual-stack node, different order, dual-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
 				{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
@@ -186,7 +186,7 @@
 			expectIPs: []net.IP{net.ParseIP("1.2.3.4"), net.ParseIP("a:b::c:d")},
 		},
 		{
-			name:      "dual-stack node, IPv6-first, no internal IPv4, single-stack cluster",
+			name: "dual-stack node, IPv6-first, no internal IPv4, single-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
 				{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
@@ -196,7 +196,7 @@
 			expectIPs: []net.IP{net.ParseIP("a:b::c:d")},
 		},
 		{
-			name:      "dual-stack node, IPv6-first, no internal IPv4, dual-stack cluster",
+			name: "dual-stack node, IPv6-first, no internal IPv4, dual-stack cluster",
 			addresses: []v1.NodeAddress{
 				{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
 				{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
@@ -209,7 +209,7 @@
 	}
 
 	for _, tc := range testcases {
-		t.Run(tc.name, func (t *testing.T) {
+		t.Run(tc.name, func(t *testing.T) {
 			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, tc.dualStack)()
 			node := &v1.Node{
 				Status: v1.NodeStatus{Addresses: tc.addresses},

Run ./hack/update-gofmt.sh
+++ exit code: 1
+++ error: 1
[0;31mFAILED[0m   verify-gofmt.sh	16s
Verifying verify-golint.sh

+++ Running case: verify.golint 
+++ working dir: /home/prow/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-golint.sh"
installing golint
go: downloading golang.org/x/lint v0.0.0-20200302205851-738671d3881b
go: downloading golang.org/x/tools v0.0.0-20200710042808-f1c4188a97a1
Errors from golint:
pkg/kubelet/volume_host.go:237:9: if block ends with a return statement, so drop this else and outdent its block

Please review the above warnings. You can test via "golint" and commit the result.
If the above warnings do not make sense, you can exempt this package from golint
checking by adding it to hack/.golint_failures (if your reviewer is okay with it).

+++ exit code: 1
+++ error: 1
[0;31mFAILED[0m   verify-golint.sh	136s
```

andrewsykim commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -262,23 +262,23 @@ func (kl *Kubelet) GetPodCgroupRoot() string {
-// GetHostIP returns host IP or nil in case of error.
-func (kl *Kubelet) GetHostIP() (net.IP, error) {
+// GetHostIPs returns host IPs or nil in case of error.
+func (kl *Kubelet) GetHostIPs() ([]net.IP, error) {
```
Since callers of this always extract the IP at index 0, would it be better to keep this as GetHostIP and and return index 0 from the returned list at GetNodeHostIPs?

danwinship replied:
I guess there's only one caller... I don't know, it seemed more appropriate to me to have the function return all the relevant information, and let the caller discard the secondary IP if it doesn't need it...

andrewsykim replied:
Okay good point, missed that there is only 1 caller.

andrewsykim commented on pkg/kubelet/nodestatus/setters_test.go:
```diff
@@ -443,6 +443,74 @@ func TestNodeAddress(t *testing.T) {
 }
 
+// We can't test failure or autodetection cases here because the relevant code isn't mockable
+func TestNodeAddress_BareMetal(t *testing.T) {
```
nit: `TestNodeAddress_WithoutCloudProvider` might be a bit more accurate naming for this test since bare metal can technically have a "cloud provider" enabled

andrewsykim commented on pkg/kubelet/volume_host.go:
```diff
@@ -231,7 +231,12 @@ func (kvh *kubeletVolumeHost) GetHostName() string {
+	hostIPs, err := kvh.kubelet.GetHostIPs()
+	if err != nil {
+		return nil, err
+	} else {
```
I think this else is redundant?

danwinship commented:
/retest

khenidak commented on cmd/kubelet/app/options/options.go:
```diff
@@ -325,7 +325,7 @@ func (f *KubeletFlags) AddFlags(mainfs *pflag.FlagSet) {
 	fs.StringVar(&f.HostnameOverride, "hostname-override", f.HostnameOverride, "If non-empty, will use this string as identification instead of the actual hostname. If --cloud-provider is set, the cloud provider determines the name of the node (consult cloud provider documentation to determine if and how the hostname is used).")
 
-	fs.StringVar(&f.NodeIP, "node-ip", f.NodeIP, "IP address of the node. If set, kubelet will use this IP address for the node. If unset, kubelet will use the node's default IPv4 address, if any, or its default IPv6 address if it has no IPv4 addresses. You can pass '::' to make it prefer the default IPv6 address rather than the default IPv4 address.")
+	fs.StringVar(&f.NodeIP, "node-ip", f.NodeIP, "IP address (or comma-separated dual-stack IP addresses) of the node. If unset, kubelet will use the node's default IPv4 address, if any, or its default IPv6 address if it has no IPv4 addresses. You can pass '::' to make it prefer the default IPv6 address rather than the default IPv4 address.")
```
the `if set` part of the doc has been removed. Is this intentional?

danwinship replied:
It seemed like it would be awkward to rewrite the sentence in a way that reflected the fact that the value could be one or two addresses, and it seemed kind of obvious and unnecessary to state anyway... ("kubelet --node-ip tells kubelet to set the node IP")

aojea replied:
so, in order to have a kubelet with dual stack addresses you always have to pass always the node-ips as a parameter.
This is a pita for installation automation :/ but since you did an extensive research on this topic I imagine that is not easy to support/add it

aojea replied:
ic, RTFM thing , it is in the KEP

khenidak commented on pkg/kubelet/kubelet_network_linux.go:
```diff
@@ -23,15 +23,39 @@ import (
+	maybeDualStack := utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack)
+	ipv6Primary := kl.nodeIPs != nil && utilnet.IsIPv6(kl.nodeIPs[0])
+
+	var iptClients []utiliptables.Interface
```
this `iptCllient` is the same as one defined on Kubelet `kl.iptClient`? did we remove it from Kubelet as we now pass it around?

danwinship replied:
Yes (https://github.com/kubernetes/kubernetes/pull/95239/commits/8a264e3a2d548a4e923c35b718c0a142050648e7#diff-bf28da68f62a8df6e99e447c4351122dL822). Nothing else in kubelet uses it, and given that the easiest way to make this code dual-stack was to call `syncNetworkUtil` twice with different clients, there was no longer any good reason to store the client(s) in the kubelet.

khenidak commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,59 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeInternalIP {
+			ip := net.ParseIP(addr.Address)
+			if ip != nil {
```
shouldn't we at least log this?

danwinship replied:
Hm... I don't like logging messages from "get" methods. Who knows how often it might be called.

This is mostly a "can't happen". (It can only happen if you're using a broken cloud provider. But apparently someone was, because there's a bug report.) I guess kubelet should log an error if it observes the cloud provider setting an invalid IP from the node-address-setting loop.

khenidak commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,59 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeExternalIP {
+			ip := net.ParseIP(addr.Address)
+			if ip != nil {
```
similar to previous comment re logging failures

khenidak commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,59 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
+		return nil, fmt.Errorf("host IP unknown; known addresses: %v", node.Status.Addresses)
+	}
+
+	nodeIPs := []net.IP{allIPs[0]}
```
Order is assumed here. Which is fine if we know that the node.Status.Addresses is always ordered correctly. is this confirmed?

danwinship replied:
This function is what defines what "ordered correctly" is for node.Status.Addresses. Whatever IP the cloud provider returns first is the one that kubelet will use.

Note that the first IP returned here is the same as was traditionally returned by the old single-value `GetNodeHostIP()`; "the first InternalIP, unless there are no InternalIPs, in which case the first ExternalIP". I just changed the way it figures that out because with dual-stack we need to do that logic twice so it's easier to just sort the IPs correctly beforehand.

khenidak commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,59 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
+			}
+		}
+	}
+	for _, addr := range node.Status.Addresses {
```
why two loops?

danwinship replied:
The first loop gets the InternalIPs, the second gets the ExternalIPs.

khenidak commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,59 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
+		return nil, err
 	}
-	return nil, fmt.Errorf("host IP unknown; known addresses: %v", addresses)
+	return ips[0], nil
```
I understand that `GetNodeHostIPs(..)` will return error in case of len()==0. But a comment here will make it clearer.

khenidak commented on pkg/util/node/node_test.go:
```diff
@@ -89,6 +94,147 @@ func TestGetPreferredAddress(t *testing.T) {
+				if tc.expectIPs != nil {
+					t.Errorf("expected %v, got error (%v)", tc.expectIPs, err)
+				}
+			} else if tc.expectIPs == nil {
```
that does not test that func returned error. Just checks for len of slice.

danwinship replied:
The test case knows that GetNodeHostIPs returns an error XOR a non-empty slice, so the testcase struct does not have separate `expectIPs` and `expectError` fields. At this point we know that `err` is `nil` and `tc.expectIPs` is empty, so that means we expected an error and didn't get one, as the error message says.

So I think this is correct. If you want something changed please clarify what part of it

khenidak commented on pkg/util/node/BUILD:
```diff
@@ -11,15 +11,18 @@ go_library(
     srcs = ["node.go"],
     importpath = "k8s.io/kubernetes/pkg/util/node",
     deps = [
+        "//pkg/features:go_default_library",
```
low pri - Would be great if we move all the generated stuff into a separate commit, making it easier to skip in review

khenidak commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -57,25 +57,46 @@ const (
 
+	if len(nodeIPs) > 1 {
+		if externalCloudProvider || cloud != nil {
+			klog.Warningf("Cloud providers do not support multiple --node-ip values; ignoring second value")
```
that is not entirely true. That means that cloud hosted clusters will always have host net with single IP? 

can we double check @andrewsykim would know more.

danwinship replied:
I worded this badly. It's not supposed to be saying anything about what the cloud provider can or can't do, it's just saying that _kubelet_ does not allow you to pass multiple `--node-ip` values if you are using a cloud provider, at this time.

(I assume we will eventually allow dual `--node-ip` with clouds as well, but this is more complicated because, among other things, the behavior of `--node-ip` is currently inconsistent between internal cloud providers and external cloud providers. This is one of the things discussed in KEP-1664 that isn't resolved yet.)

(So, the behavior both before and after this PR is that a cloud provider can return dual-stack node addresses, and kubelet will allow you to override _one_ of them, but not both.)

khenidak commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -57,25 +57,46 @@ const (
+		if externalCloudProvider || cloud != nil {
+			klog.Warningf("Cloud providers do not support multiple --node-ip values; ignoring second value")
+		} else if nodeIP.IsUnspecified() || secondaryNodeIP.IsUnspecified() {
+			klog.Warningf("Multi-valued --node-ip cannot include 0.0.0.0 or ::; ignoring second value")
```
this needs to move to option validation in option file. if we are to resolve the cloud comment earlier then we need to leave this validation here.

khenidak commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1531,14 +1531,17 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
-				s.PodIP = hostIP.String()
+				s.PodIP = hostIPs[0].String()
 				s.PodIPs = []v1.PodIP{{IP: s.PodIP}}
+				if len(hostIPs) == 2 {
```
can we add a test for this?

danwinship replied:
done

khenidak reviewed (changes requested):
few issues. The most concerning is the order of IPs dependancy.

danwinship commented on pkg/kubelet/nodestatus/setters_test.go:
```diff
@@ -443,6 +443,74 @@ func TestNodeAddress(t *testing.T) {
+			}
+
+			// Sort both sets for consistent equality
+			sortNodeAddresses(testCase.expectedAddresses)
```
huh... this was just copied from the existing `TestNodeAddress`...

ugh, the existing test is broken. There are a mix of "make sure addresses X gets filtered out of the results" tests and "make sure address X gets moved to the front of the list" tests, but the latter set are all no-ops. :facepalm:

danwinship replied:
fixed

danwinship committed:
kubelet: Remove unnecessary sorting in NodeAddress tests

Several of the tests in TestNodeAddress() were no-ops because the test
code was only testing that NodeAddresses() returned all of the
expected addresses, but not testing that it was returning them in the
correct order.

The order that NodeAddresses() returns addresses in is very important,
so fix the tests to actually test it.

One existing test ("NodeIP is external") had its expectedAddresses in
the wrong order, but it seems clear from the name of the test that
this isn't actually what it expected.

Also, previously testKubeletHostname was "127.0.0.1" which ended up
interacting weirdly with the IPv4-vs-IPv6 sorting code in a way that
made some of the test results confusing if you didn't realize that
testKubeletHostname was an IPv4 address. Fix that by making it an
actual hostname instead, which then preserves the expected sorting.

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -462,7 +463,19 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
+	for _, nodeIP := range strings.Split(nodeIP, ",") {
+		parsedNodeIP := net.ParseIP(strings.TrimSpace(nodeIP))
+		if parsedNodeIP == nil {
+			klog.Warningf("Could not parse --node-ip value %q; ignoring", nodeIP)
```
Is it right to ignore this with no other signals?  I am not sure what is best but this seems likely to get lost in the noise

danwinship replied:
Well, currently we ignore invalid `--node-ip` _without_ logging a warning, so this is at least an improvement...

Assuming we add `--node-ips` or `--node-ip-mode` in KEP-1664, we can make errors fatal there...

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -462,7 +463,19 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
+			nodeIPs = append(nodeIPs, parsedNodeIP)
+		}
+	}
+	if len(nodeIPs) > 2 || (len(nodeIPs) == 2 && utilnet.IsIPv6(nodeIPs[0]) == utilnet.IsIPv6(nodeIPs[1])) {
```
Does this need to be gated?

danwinship replied:
Define "need"...

There is no chance that a user is going to be accidentally injured by us not gating it, since they can only end up with two IPs here if they explicitly passed two IPs on the command line.

Note that the eventual side effect of setting dual-stack `podIPs` _is_ gated.

thockin commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -262,23 +262,23 @@ func (kl *Kubelet) GetPodCgroupRoot() string {
 		return nil, fmt.Errorf("cannot get node: %v", err)
 	}
-	return utilnode.GetNodeHostIP(node)
+	return utilnode.GetNodeHostIPs(node)
```
DO we now need to pluralize pod.status.hostIP ?

aojea replied:
there is an open issue about that https://github.com/kubernetes/kubernetes/issues/85443
It seems so

danwinship replied:
I mentioned that in the KEP; it's not a critical fix, since for the most part, every client in a dual-stack cluster ought to be able to reach either node IP, so having easy API access to one of them but not the other should not be fatal. But probably should be fixed for beta.

thockin reviewed (commented):
I am approving, @khenidak owns the LGTM

/approve

khenidak commented:
One point re validation. But can be added later.

/LGTM

danwinship commented:
@khenidak ah, I was in the middle of updating when you lgtm'ed; I addressed your comment about where node-ip validation happens. See if this is better.

/hold
need to rebase into openshift to test

khenidak commented:
@danwinship great thank you. i will keep an eye until the hold is removed.

danwinship commented:
/hold cancel
(had to repush to fix bazel BUILD files)

aojea commented:
/test pull-kubernetes-e2e-kind-dual-canary
/retest

aojea commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -373,11 +374,15 @@ func (c *Configurer) GetPodDNS(pod *v1.Pod) (*runtimeapi.DNSConfig, error) {
-			case c.nodeIP.To16() != nil:
-				dnsConfig.Servers = []string{"::1"}
+			for _, nodeIP := range c.nodeIPs {
+				if utilnet.IsIPv6(nodeIP) {
+					dnsConfig.Servers = append(dnsConfig.Servers, "::1")
+				} else {
+					dnsConfig.Servers = append(dnsConfig.Servers, "127.0.0.1")
```
I have to confess that this part scares me, how this is going to behave with both nameservers.
maybe just adding `localhost` and let the host handle it?

aojea commented:
/hold
need to check something, sorry

aojea commented on pkg/util/node/node.go:
```diff
@@ -90,27 +93,60 @@ func GetPreferredNodeAddress(node *v1.Node, preferredAddressTypes []v1.NodeAddre
-func GetNodeIP(client clientset.Interface, hostname string) net.IP {
+// GetNodeIP returns an IP (as with GetNodeHostIP) for the node with the provided name.
+// If required, it will wait for the node to be created.
+func GetNodeIP(client clientset.Interface, name string) net.IP {
```
ok, I've found it, this has a dependency in kube-proxy but does not break it
https://github.com/kubernetes/kubernetes/blob/aa1f91286838d373a6d9217ad9f4fe0647a1c859/cmd/kube-proxy/app/server_others.go#L136-L142

possible something to follow up or do you want to fix it here?
anyway, unhold as you wish, I just was worried about this dependency in kube-proxy

aojea commented:
/test pull-kubernetes-e2e-kind-dual-canary
/retest

aojea commented:
/test pull-kubernetes-e2e-gce-alpha-features

khenidak commented:
/test pull-kubernetes-e2e-kind-dual-canary

khenidak commented:
@danwinship + @aojea  according to my understanding this test "pull-kubernetes-e2e-kind-dual-canary " should start working with this PR? (@aojea: do we need to add anything to kind, e.g. making kubelet report two IPs?)

aojea commented:
> we need to add anything to kind, e.g. making kubelet report two IPs?)

yep,I need to modify kind , we need to pass the node ips to  kubelet using the --node-ip ipv4,ipv6 format.

khenidak commented:
> > we need to add anything to kind, e.g. making kubelet report two IPs?)
> 
> yep,I need to modify kind , we need to pass the node ips to kubelet using the --node-ip ipv4,ipv6 format.

Thanks.  let us move this PR forward and get the kind dual canary adjusted, so we can use it for everything dualstack..

khenidak commented:
/lgtm

khenidak commented:
/retest

aojea commented:
/retest
jobs timing out

aojea commented:
/retest
job timeouts

dcbw commented:
more job times `Job execution failed: Pod pending timeout.`
/retest

khenidak commented:
@dcbw it is across the board for multiple PRs. not sure if it is github or prow.

aojea commented:
> @dcbw it is across the board for multiple PRs. not sure if it is github or prow.

prow 

/retest
